Modules/_sqlite/cursor.c

```diff
@@ -121,6 +116,14 @@ cursor_dealloc(pysqlite_Cursor *self)
     if (self->in_weakreflist != NULL) {
         PyObject_ClearWeakRefs((PyObject*)self);
     }
+    if (self->statement) {
+        /* A SELECT query will lock the affected database table(s), so we need
+         * to reset the statement to unlock the database before disappearing */
+        sqlite3_stmt *stmt = self->statement->st;
+        if (sqlite3_stmt_readonly(stmt)) {
+            pysqlite_statement_reset(self->statement);
+        }
+    }
```

Comment:
Hummm, I am not completely sure if this is correct. The docs mention:

> Note that application-defined SQL functions or virtual tables might change the database indirectly as a side effect. >For example, if an application defines a function "eval()" that calls sqlite3_exec(), then the following SQL statement would change the database file through side-effects:
> `SELECT eval('DELETE FROM t1') FROM t2;`
> But because the SELECT statement does not change the database file directly, sqlite3_stmt_readonly() would still return true.

So I think `sqlite3_stmt_readonly` may give false positives but I am not sure if you have something more in mind that I may be missing.

Reply:
Yes, that's true, but AFAICS, we can't really protect ourselves against such (mis)use :/

Reply:
Well, we _could_ of course always reset when a statement exits from the cursor (dealloc or being replaced by a new statement). But this is a weird corner case.

Reply:
> Well, we _could_ of course always reset when a statement exits from the cursor

That's what I am referring to, it seems to me that if I read this correctly, that's what the old code used to do, no?

Reply:
Well, the old code resets all the time, so yes that is true. However, I do find that example a bit theoretical. That does not necessarily mean that we should not protect ourselves from it, though.

Reply:
I thought this over once more. TL;DR: I suggest keeping it like it is :)

We want to reset `SELECT` queries, and why is that? Because other queries are "single step"; that is, they're done after calling `sqlite3_step` once. `SELECT` queries may return rows, so the statement is kept "alive" (and the affected table(s) locked) until there is no more rows to return. Thus, we need the reset when the current statement "exits" the cursor. In order to make sure we reset all `SELECT` queries, we use `sqlite3_stmt_readonly`, which return true for all `SELECT` queries. So it does not matter if the query uses some fancy UDF that also alters the database; we still want to reset the statement, and we will, because `sqlite3_stmt_readonly` still return true.

If `sqlite3_stmt_readonly` starts behaving differently, we can remove the check, but I do not think that will happen, given SQLite's strict backwards compatibility policy. What we could (and should) do, though, is to add a test for this, using something like the example from the SQLite docs.

Reply:
I am convinced 👍